scalyr_agent/copying_manager/worker.py

```diff
@@ -1215,7 +1215,83 @@ def create_shared_object_manager(worker_class, worker_proxy_class):
+        # we need to be sure that the worker is stopped first.
+        # This class attribute should provide access to the worker for the watchdog that is implemented below.
+        worker = None
+
+        @classmethod
+        def _create_worker(cls, *args, **kwargs):
+            """
+            Create a new worker and save it as a 'cls.worker' class attribute
+            to be able to access the worker's instance by the watchdog.
+            :return: the proxy object for the 'cls.worker'
+            """
+            cls.worker = worker_class(*args, **kwargs)
+            return cls.worker
+
+        @classmethod
+        def _run_server(cls, *args, **kwargs):
+            """
+            This is the entry point of the SyncManager's process
+            and it is overridden to handle the situation where parent process was killed.
+            The current implementation of the multiprocess.managers does not provide ability to detect such situation
+            and the orphan process of the SyncManager still continues working even after the parent was killed.
+
+
+            To achieve needed behaviour, before we call the original '_run_server', we create
+            and start a 'watchdog' thread, which checks the existence of the parent process
+            and terminates everything after the parent dies.
+            :param args: passed directly to the original parent method.
+            :param kwargs: passed directly to the original parent method.
+            """
+
+            def run_watchdog():
+                """
+                Periodically check the parent process for existence and terminate everything if process is not found.
+                :return:
+                """
+                while True:
+                    try:
+                        # probing the parent process.
+                        if cls.parent_pid is not None:
+                            os.kill(cls.parent_pid, 0)
+                        time.sleep(1)
+                    except OSError:
+                        # parent is not found.
+                        # stop the worker if it is still alive.
+                        if cls.worker and cls.worker.is_alive():
+                            try:
+                                # Try to stop worker gracefully.
+                                # If this unsuccessful, then just ignore that
+                                # because the whole process will be killed later.
+                                cls.worker.stop_worker()
+                            except:
+                                # ignore if worker can not be stopped everything will be killed later anyway.
+                                pass
+                            finally:
+                                break
+
+                # try to terminate everything gracefully first.
+                os.kill(os.getpid(), signal.SIGTERM)
+                time.sleep(1)
```

Comment:
This should be greater than 1.  I would suggest 5 seconds.  Also, replace with the `delay_before_kill` variable.